### DISCUSSION TITLE
The ability to have protoset templates

### DISCUSSION THREAD

**Discussion by a-magno** [peter-kish/gloot#95] [Ideas]
As the title implies, it would speed up the creation of large quantities of item protosets if we could, say, save a certain protoset as a "base" with all item attributes needed, and then load that base into a new item protoset through, maybe, the file loading system native to Godot.

Going into a step-by-step explanation, it could work something like this:
<ol>
<li>User makes an item protoset as usual, listing all the desired attributes items should have (width, height, image, value, name, description, etc...)</li>
<li>Then, the user could, through means of a "Save Template" button, save that list of attributes as a resource with a dictionary of attribute_name as a key, and type_hint as the value (assuming you don't want templates to hold default values).</li>
<li>Now, when the user makes a new item, they can click a "Load From Template" button, which opens a File Open popup, so they can select the template resource file.</li>
<li>Once the file has been loaded and unpacked, add all the attributes within the dictionary into the newly created protoset, where the user can now edit their values to their liking.</li>
</ol>

I believe this could speed up the usage of the plugin considerably, and I have been trying to make my own implementation of this idea as well, but thought it would be interesting to suggest and see the author's own take on it!

**Comment by peter-kish** [comment]
Hi,
So, I suppose the idea is to speed up the creation of new protosets by using templates?
This could be done about how you described, but I don't see a lot of purpose in producing a large number of protosets for a single project. Actually, I'd be surprised if you needed more than one protoset per project as I intended it to contain all the item prototypes you would need.

Unless by "creating of large quantities of item protosets" you meant "creating a large amount of item prototypes", in which case the whole idea makes a lot more sense. In that case having prototype templates could speed things up if you plan to create a lot of item prototypes. An alternative could also be to just be able to duplicate already existing prototypes and modify them to create new variations (this could be the first implementation as it's much simpler).

Unfortunately, I still didn't have the chance to use my plugin on a complex project, but speeding up the process of populating the protoset with item prototypes makes a lot of sense as some projects could easily contain hundreds of different items. that would need to be defined individually.

**Reply by a-magno** [reply]
Hello,
Yes, I meant prototypes, I got the terminologies mixed up, since I started using GLoot only recently. 

Being able to duplicate existing prototypes is definitely the simpler implementation for the state of the plugin right now, I could even see how it would be implemented, along with the option to, say, duplicate a property within a prototype (sometimes I catch myself wishing I could do it). 

If it's alright with you, I could experiment with adding the templating functionality to the addon, as I have started to curiously poke around to get a feel for how to best use it in my projects, and I believe it shouldn't impact the code as it is. Right now, this is what I'm thinking of, for basic logic:

1. Add button somewhere near the prototype list for duplicating.
2. Button, when clicked, makes a new item prototype
3. Selected prototype has it's data copied into the new prototype, probably changes the name to avoid conflict with the original, something like item_prototype(1)
4. Copy is then added into the list, maybe even opened for quick editing.

I'm still a beginner when it comes to Godot, but I hope my suggestion can contribute to this plugin, since it implements my favorite style of inventory (I really like grid inventories and I think yours is the cleanest implementation I've seen yet)

**Reply by peter-kish** [reply]
Sure, feel free to experiment with the source code. The steps you described sound like a good approach. Open a pull request once you have something functional and I'll be glad to review it.

**Comment by Naeta2** [comment]
is this idea abandonned ? :(

**Reply by peter-kish** [reply]
No, I still think it would be a useful feature, it's just that development has been kinda slow lately (OMG, 6 months have already passed).

I was thinking about making support for item prototype hierarchies, similar to class inheritance in object-oriented programming. That way common properties could be grouped for a number of "derived" prototypes (I think [Pandora](https://github.com/bitbrain/pandora) has a similar approach).

Anyways, that will require a lot of changes, so maybe I'll just implement prototype duplication for the next version and leave the inheritance for a later release.
